src/Mvc/Mvc.RazorPages/src/Infrastructure/DynamicPageEndpointMatcherPolicy.cs

```diff
@@ -0,0 +1,126 @@
+
+        public DynamicPageEndpointMatcherPolicy(DynamicPageEndpointSelector selector, IPageLoader loader)
+        {
+            if (selector == null)
+            {
+                throw new ArgumentNullException(nameof(selector));
+            }
+
+            if (loader == null)
+            {
+                throw new ArgumentNullException(nameof(loader));
+            }
+
+            _selector = selector;
+            _loader = loader;
+        }
+
+        public override int Order => int.MinValue + 100;
+
+        public bool AppliesToEndpoints(IReadOnlyList<Endpoint> endpoints)
+        {
+            if (endpoints == null)
+            {
+                throw new ArgumentNullException(nameof(endpoints));
+            }
+
+            if (!ContainsDynamicEndpoints(endpoints))
+            {
+                // Dynamic page endpoints are always dynamic endpoints.
+                return false;
+            }
+
+            for (var i = 0; i < endpoints.Count; i++)
+            {
+                var metadata = endpoints[i].Metadata.GetMetadata<DynamicPageMetadata>();
+                if (metadata != null)
+                {
+                    // Found a dynamic page endpoint
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        public async Task ApplyAsync(HttpContext httpContext, EndpointSelectorContext context, CandidateSet candidates)
+        {
+            if (httpContext == null)
+            {
+                throw new ArgumentNullException(nameof(httpContext));
+            }
+
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (candidates == null)
+            {
+                throw new ArgumentNullException(nameof(candidates));
+            }
+
+            // There's no real benefit here from trying to avoid the async state machine.
+            // We only execute on nodes that contain a dynamic policy, and thus always have
+            // to await something.
+            for (var i = 0; i < candidates.Count; i++)
+            {
+                if (!candidates.IsValidCandidate(i))
+                {
+                    continue;
+                }
+
+                var endpoint = candidates[i].Endpoint;
+
+                var metadata = endpoint.Metadata.GetMetadata<DynamicPageMetadata>();
+                if (metadata == null)
+                {
+                    continue;
+                }
+
+                var matchedValues = candidates[i].Values;
+                var endpoints = await _selector.SelectEndpointsAsync(metadata.Values);
+                if (endpoints.Count == 0)
+                {
+                    // If there's no match this is a configuration error. We can't really check
+                    // during startup that the action you configured exists.
+                    throw new InvalidOperationException(
```

Comment:
Use the resx luke

Reply:
To what end? we have no plans to add localization. It's just homework.